### PR TITLE
[#117] create authentication actor 

### LINKED ISSUE #117: Create authentication interface and memory actor
### motivation
Add the authentication actor to authenticate the request.

### job
- add authentication interface under `io.github.ztkmkoo.dss.core.actor.security` package
- add authentication command interface under `io.github.ztkmkoo.dss.core.message.security` package
- add DssRestAuthenticationActor under `io.github.ztkmkoo.dss.core.actor.rest` package
  - DssRestAuthenticationActor initialize with specific users
    - user info must contains at least user name and password
    - save this username and password list in local memory
  - DssRestAuthenticationActor can handle authentication message
    - check username and password is valid
    - if valid issue an 128-bit token
    - save the issued token in local memory
  - DssRestAuthenticationActor can handle authorization message
    - check if token is valid
    - response the token validation result(simple like true false)

### PR DESCRIPTION
**Pull request by RulLu16**
Resolves #117 

I create auth actor to authenticate request.

### DssRestAuthActor 

- Get authentication request
   If user is in userList, create jwt token and return authentication response.

- Get authorization request.
   If token is in tokenList, try to verify whether this token is valid or not and return authorization response.

But in test code, I'm sorry that there are some errors. 
So I needs you guys helps.. feel free to review for any thing!

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -113,6 +113,12 @@
             <version>${lombok.version}</version>
             <scope>provided</scope>
         </dependency>
+        
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt</artifactId>
+            <version>0.9.1</version>
+        </dependency>
```
Comment (ztkmkoo) [inline pom.xml]: @RulLu16 
First of all thank you for your job.

I have some questions bout this.
- why did you add the dependency to dss parent pom.xml not dss-core pom.xml?
- i read the document in https://github.com/jwtk/jjwt, the suggest using jjwt in general java project adding dependencies like: (it is different from yours)
``` xml
<dependency>
    <groupId>io.jsonwebtoken</groupId>
    <artifactId>jjwt-api</artifactId>
    <version>0.11.2</version>
</dependency>
<dependency>
    <groupId>io.jsonwebtoken</groupId>
    <artifactId>jjwt-impl</artifactId>
    <version>0.11.2</version>
    <scope>runtime</scope>
</dependency>
<dependency>
    <groupId>io.jsonwebtoken</groupId>
    <artifactId>jjwt-jackson</artifactId> <!-- or jjwt-gson if Gson is preferred -->
    <version>0.11.2</version>
    <scope>runtime</scope>
</dependency>
```

File: dss-core/src/test/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActorTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+				 .userID("testID")
+				 .token(testToken)
+				 .valid(probe.ref())
+				 .build());
+		 
+		 DssAuthorizationCommandResponse response = probe.expectMessageClass(DssAuthorizationCommandResponse.class);
+		 
+		 assertTrue(true);
+		 assertNotNull(response);
+	     assertEquals("true", response.getValid());
```
Comment (ztkmkoo) [inline dss-core/src/test/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActorTest.java]: @RulLu16 
Hi, i found the ci was failed here. It looks when test `handlingDssAuthorizationCommandRequest` was exected first causes the test failed.

I suggest:
- ordering the tests or make sure testToken not null(in this test, get the token before you authorization)
- improve your DssRestAuthActor that handling when token is null. => it is better that response valid false when token is invald or token is null

File: pom.xml (hunk 1 above)
Comment (RulLu16) [reply pom.xml]: In my local, there no error about dependency, so I didn't know about that. 
I'll fix this. Thank you!

File: dss-core/src/test/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActorTest.java (hunk 2 above)
Comment (RulLu16) [reply dss-core/src/test/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActorTest.java]: Thank you for review! I agree with your suggestion about ordering the tests. 
And also, I found another error about signature jwts. So trying to fix it.

File: dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActor.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    
+    private Behavior<DssAuthCommand> handlingDssAuthorizationCommandRequest(DssAuthorizationCommandRequest request) {
+    	
+    	context.getLog().info("DssAuthorizationCommandRequest: {}", request);
+    	
+    	String token = request.getToken();
+    	SecretKey key = request.getKey();
+    	String valid = "false";
+    	
+    	if(tokenList.get(token).equals(key) && verifyToken(token, key)) {
```
Comment (ztkmkoo) [inline dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActor.java]: @RulLu16 
Hi,
I think if the token not contained will cause null point exception.

tokenList.get(token) ==> null object ==> call equals method for null object

File: dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActor.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    
+    private Behavior<DssAuthCommand> handlingDssAuthenticationCommandRequest(DssAuthenticationCommandRequest request) {
+    	
+    	context.getLog().info("DssAuthenticationCommandRequest: {}", request);
+    	
+    	String userID = request.getUserID();
+    	String userPassword = request.getUserPassword();
+    	String token = null;
+    	SecretKey key = Keys.secretKeyFor(SignatureAlgorithm.HS256);
```
Comment (ztkmkoo) [inline dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActor.java]: Do we need create new secret key for every token?
Can we make just one key, and sign the jwt token?

I think it works because we put the payload with user id when create the token.

Comment (RulLu16) [reply dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActor.java]: So you think just make one key in actor? I'll try it.

Comment (ztkmkoo) [reply dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActor.java]: Yap. I think one key is enough.

Comment (ztkmkoo) [reply dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActor.java]: And do not send key back.

Comment (RulLu16) [reply dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActor.java]: In payload, there is userID already, so don't need to make new secret key. Agree with your review.
It makes codes more simple. Thanks!

File: dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActor.java (hunk 3 above)
Comment (RulLu16) [reply dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestAuthActor.java]: nullPointerException also can be caused in userList. 
I'll fix them all.

**Comment by ztkmkoo** [comment]
@RulLu16 
Thanks for your contribution.
After update the branch, i will merge this pr

**Comment by RulLu16** [comment]
> @RulLu16
> Thanks for your contribution.
> After update the branch, i will merge this pr

@ztkmkoo 
Thank you for your review! Thanks to your help, I resolve this issue.
